spec/controllers/books_controller_spec.rb

```diff
@@ -1,40 +1,49 @@
 require 'rails_helper'
-
 RSpec.describe BooksController, type: :controller do
-  before do
-    get :index
-  end
-
   describe '#index' do
-    context 'empty book' do
+    before { get :index }
+    context 'has no book' do
       it 'assigns empty to books' do
         expect(assigns(:books)).to be_empty
       end
     end
 
-    context 'has one book' do
+    context 'has many books' do
       before do
-        @book = create(:book)
+        3.times do
+          user = create(:user)
+          3.times do
+            create(:book, user_id: user.id)
+          end
+        end
       end
 
-      it 'assigns empty to books' do
-        expect(assigns(:books).count).to equal(1)
-        expect(assigns(:books).first.id).to equal(@book.id)
+      it 'assigns many book to books' do
+        expect(assigns(:books).count).to equal(Book.count)
+        expect(assigns(:books).pluck(:id)).to match_array(Book.all.pluck(:id))
       end
     end
+  end
 
-    context 'has many books' do
-      before do
-        @book_ids = []
-        5.times do
-          book = create(:book)
-          @book_ids << book.id
-        end
+  describe '#show' do
+    before do
+      @book = create(:book)
+      3.times do
+        create(:image, book_id: @book.id)
+      end
+    end
+    context 'find result' do
+      it 'book found' do
+        get :show, params: { id: @book.id }
+        expect(assigns(:book).attributes).to eql(@book.attributes)
       end
 
-      it 'assigns empty to books' do
-        expect(assigns(:books).count).to equal(@book_ids.count)
-        expect(assigns(:books).pluck(:id)).to match_array(@book_ids)
+      it 'book not found' do
+        allow(Book).to receive(:find_by).with(anything).and_return(nil)
+        get :show, params: { id: @book.id }
```

Comment:
Same as line 37.

Reply:
Nếu before thì get :show, params: { id: @book.id } sẽ chạy trước allow(Book).to receive(:find_by).with(anything).and_return(nil)

Reply:
Ok.